server/core_benchmarks_test.go

```diff
@@ -249,3 +251,317 @@ func BenchmarkCoreTLSFanOut(b *testing.B) {
+								quitChs[i] = make(chan bool, 1)
+							}
+
+							// TODO rename
+							completedPublishersCount := 0
+
+							var benchCompleteWg sync.WaitGroup
+							benchCompleteWg.Add(1)
+
+							ncSub.Subscribe(fmt.Sprintf("%s.*", subjectBaseName), func(msg *nats.Msg) {
+								// split subject by "." and get the publisher id
+								pubIdx, err := strconv.Atoi(strings.Split(msg.Subject, ".")[1])
+								if err != nil {
+									b.Fatal(err)
+								}
+
+								counters[pubIdx] += 1
+								if counters[pubIdx] == b.N {
+									completedPublishersCount++
+									if completedPublishersCount == numPubs {
+										benchCompleteWg.Done()
+									}
+								}
+							})
+
+							// random bytes as payload
+							messageData := make([]byte, messageSize)
+							rand.Read(messageData)
+
+							var publishersReadyWg sync.WaitGroup
+							// waits for all publishers sub-routines and for main thread to be ready
+							publishersReadyWg.Add(numPubs + 1)
+
+							// wait group to ensure all publishers have been torn down
+							var finishedPublishersWg sync.WaitGroup
+							finishedPublishersWg.Add(numPubs)
+
+							// create N publishers
+							for i := 0; i < numPubs; i++ {
+								// create publisher connection
+								ncPub, err := nats.Connect(clientUrl, opts...)
+								if err != nil {
+									b.Fatal(err)
+								}
+								defer ncPub.Close()
+
+								go func(pubId int) {
+									// signal that this publisher has been torn down
+									defer finishedPublishersWg.Done()
+
+									subject := fmt.Sprintf("%s.%d", subjectBaseName, pubId)
+
+									// publisher successfully initialized
+									publishersReadyWg.Done()
+
+									// wait till all other publishers are ready to start workload
+									publishersReadyWg.Wait()
+
+									// publish until quitCh is closed
+									for {
+										select {
+										case <-quitChs[pubId]:
+											return
+										default:
+											// continue publishing
+										}
+										err := ncPub.Publish(subject, messageData)
+										if err != nil {
+											errors[pubId] += 1
+										}
+									}
+								}(i)
+							}
+
+							// Set bytes per operation
+							b.SetBytes(messageSize)
+							// Main thread is ready
+							publishersReadyWg.Done()
+							// Wait till publishers are ready
+							publishersReadyWg.Wait()
+
+							// Start the clock
+							b.ResetTimer()
+							// wait till termination cond reached
+							benchCompleteWg.Wait()
+							// Stop the clock
+							b.StopTimer()
+
+							// send quit signal to all publishers
+							for pubIdx := range quitChs {
+								quitChs[pubIdx] <- true
+							}
+							// Wait for all publishers to shutdown
+							finishedPublishersWg.Wait()
+
+							// sum errors from all publishers
+							totalErrors := 0
+							for _, err := range errors {
+								totalErrors += err
+							}
+
+							// report errors
+							b.ReportMetric(float64(totalErrors), "errors")
```

Comment:
Same as above, rate would be more informative, this is just a number that would be hard to make sense of.